src/haz3lcore/dynamics/elaborator.re

```diff
@@ -51,222 +51,237 @@ let exp_binop_of: Term.UExp.op_bin => (Typ.t, (_, _) => DHExp.t) =
       ((e1, e2) => BinStringOp(string_op_of(op), e1, e2)),
     );
 
+/* Wrap: Handles cast insertion and non-empty-hole wrapping
+   for elaborated expressions */
+let wrap = (u, mode, self, d: DHExp.t): option(DHExp.t) =>
+  switch (Statics.error_status(mode, self)) {
+  | NotInHole(_) =>
+    switch (mode) {
+    | Syn => Some(d)
+    | SynFun =>
+      /* Things in function position get cast to an arrow type */
+      let ty_self = Typ.t_of_self(self);
+      switch (ty_self) {
+      | Unknown(prov) =>
+        Some(DHExp.cast(d, ty_self, Arrow(Unknown(prov), Unknown(prov))))
+      | _ => Some(d)
+      };
+    | Ana(ana_ty) =>
+      /* Forms with special ana rules get cast from their appropriate Matched types */
+      switch (d) {
+      | ListLit(_)
+      | Cons(_) =>
+        switch (ana_ty) {
+        | Unknown(prov) =>
+          Some(DHExp.cast(d, List(Unknown(prov)), ana_ty))
+        | _ => Some(d)
+        }
+      | Fun(_) =>
+        switch (ana_ty) {
+        | Unknown(prov) =>
+          Some(DHExp.cast(d, Arrow(Unknown(prov), Unknown(prov)), ana_ty))
+        | _ => Some(d)
+        }
+      /*| Tuple(_) =>
+        switch (ana_ty) {
+        | Prod(_) => Some(d)
+        | _ => Some(DHExp.cast(d, Typ.t_of_self(self), ana_ty))
+        }*/
+      | Tuple(ds) =>
+        switch (ana_ty) {
+        | Unknown(prov) =>
+          let us = List.init(List.length(ds), _ => Typ.Unknown(prov));
+          Some(DHExp.cast(d, Prod(us), ana_ty));
+        | _ => Some(d)
+        }
+      | Inj(_) =>
+        switch (ana_ty) {
+        | Unknown(prov) =>
+          Some(DHExp.cast(d, Sum(Unknown(prov), Unknown(prov)), ana_ty))
+        | _ => Some(d)
+        }
+      | FixF(_) =>
```

Comment:
I don't think this case ever arises but in principle, it should behave like Let in that the expected type is passed down in the statics so I'd put it down in the next group of cases